test/syscalls/linux/mq.cc

```diff
@@ -0,0 +1,283 @@
+// Copyright 2021 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <fcntl.h>
+#include <mqueue.h>
+#include <sched.h>
+#include <sys/poll.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <string>
+
+#include "test/util/capability_util.h"
+#include "test/util/posix_error.h"
+#include "test/util/test_util.h"
+
+#define NAME_MAX 255
+
+namespace gvisor {
+namespace testing {
+namespace {
+
+// PosixQueue is a RAII class used to automatically clean POSIX message queues.
+class PosixQueue {
+ public:
+  PosixQueue(mqd_t fd, std::string name) : fd_(fd), name_(std::string(name)) {}
+
+  // Move constructor.
+  PosixQueue(PosixQueue&& q) : fd_(q.fd_), name_(q.name_) {
+    // Call PosixQueue::release, to prevent the object being released from
+    // unlinking the underlying queue.
+    q.release();
+  }
+
+  ~PosixQueue() {
+    if (fd_ != -1) {
+      EXPECT_THAT(mq_close(fd_), SyscallSucceeds());
+      EXPECT_THAT(mq_unlink(name_.c_str()), SyscallSucceeds());
+    }
+  }
+
+  mqd_t fd() { return fd_; }
+
+  const char* name() { return name_.c_str(); }
+
+  mqd_t release() {
+    mqd_t old = fd_;
+    fd_ = -1;
+    return old;
+  }
+
+ private:
+  mqd_t fd_;
+  std::string name_;
+};
+
+// genName generates an unused name to be used by mq_open(2) to create message
+// queues. genName generates the name by choosing a psuedo-random name, and
+// testing if it exists. While it's most likely that a name will be found in
+// the first few iterations, it is possible that genName may loop infinitely.
+std::string genName() {
```

Comment:
Done. I removed `genName`, exported `NextTempBasename`, and used it instead. If you meant that we should use `NextTempBasename` inside `genName` not instead of it, let me know and I'll update it.